### ISSUE TITLE
js: nested template highlighting

### ISSUE THREAD

**Issue by aeosynth** [mawww/kakoune#1670]
javascript template literals can have more template literals embedded within them. kak does not highlight nested templates correctly: 

```js
`string ${`also a string: let x = 42`}`
```

![js](https://user-images.githubusercontent.com/75317/32411839-ec532aa2-c1a2-11e7-8f97-1f3551b314bc.png)

**Comment by danr** [comment]
I am not sure if this is possible to express given the way nested highlighters
are specified in Kakoune today. The problem is that the region highlighter's line

```
   literal       "`"  (?<!\\)(\\\\)*`         '' \
```

will exit the literal at the first backtick even if it is nested inside a `${...}`:

```
`string ${`also a string: let x = 42`}`
^ start   ^ end
```

I don't see any way to tell the region highlighter to understand nestings of `${...}`.

**Comment by mawww** [comment]
Yeah, I dont think we can parse such nested syntax with the existing infrastructure, we would need more complex region definitions, where regions can contain subregions and those subregions might have the same delimiters as their parent region...

This would be quite a change in the region highlighter design, but it might make sense to do it at some point. I must admit I am not entirely satisfied with the way we define regions at the moment, it would be nice to have each region defined by its own `add-higlighter` command adding that region definition inside the region highlighter instead of hardcoding all regions in a single command.

**Comment by andychu** [comment]
FWIW I wrote an article about how nested strings are highlighted in Vim -- it has **recursive** `syntax cluster contains=`

which could be used to test any Kakoune solution:

*Three Algorithms for YSH Syntax Highlighting* - https://github.com/oils-for-unix/oils.vim/blob/main/doc/algorithms.md

https://github.com/oils-for-unix/oils.vim/blob/main/doc/stage2.md - screenshots of FIXING the nested double quote problem
